libraries/core_libs/network/src/tarcap/packets_handlers/pbft_sync_packet_handler.cpp

```diff
@@ -91,49 +99,67 @@ void PbftSyncPacketHandler::process(const PacketData &packet_data, const std::sh
     return;
   }
 
-  if (sync_block.pbft_blk->getPeriod() != pbft_mgr_->pbftSyncingPeriod() + 1) {
-    LOG(log_wr_) << "Block " << pbft_blk_hash << " period unexpected: " << sync_block.pbft_blk->getPeriod()
+  if (period_data.pbft_blk->getPeriod() != pbft_mgr_->pbftSyncingPeriod() + 1) {
+    LOG(log_wr_) << "Block " << pbft_blk_hash << " period unexpected: " << period_data.pbft_blk->getPeriod()
                  << ". Expected period: " << pbft_mgr_->pbftSyncingPeriod() + 1;
     restartSyncingPbft(true);
     return;
   }
 
-  // Check cert vote matches
-  for (auto const &vote : sync_block.cert_votes) {
-    if (vote->getBlockHash() != pbft_blk_hash) {
-      LOG(log_er_) << "Invalid cert votes block hash " << vote->getBlockHash() << " instead of " << pbft_blk_hash
-                   << " from peer " << packet_data.from_node_id_.abridged() << " received, stop syncing.";
-      handleMaliciousSyncPeer(packet_data.from_node_id_);
-      return;
+  // Check cert vote matches if final synced block
+  if (pbft_chain_synced) {
+    for (auto const &vote : current_block_cert_votes) {
+      if (vote->getBlockHash() != pbft_blk_hash) {
+        LOG(log_er_) << "Invalid cert votes block hash " << vote->getBlockHash() << " instead of " << pbft_blk_hash
+                     << " from peer " << packet_data.from_node_id_.abridged() << " received, stop syncing.";
+        handleMaliciousSyncPeer(packet_data.from_node_id_);
+        return;
+      }
+    }
+  }
+
+  // Check votes match the hash of previous block in the queue
+  auto last_pbft_block = pbft_mgr_->periodDataQueueLastPbftBlock();
+  if (last_pbft_block) {
```

Comment:
this is not going to validate the hashes in many cases, which is not ok. The thing is, we have to have a way to validate these votes hashes every single time we receive sync data. Anytime you start syncing and get first block, we will skip validation due to this condition, anytime the processing thread that is popping data from sync queue faster than we are pushing it there (as a result queue is almost always empty) we will skip validation... We have to save somehow somewhere last processed pbft_block hash and this validation must be run every single time we receive sync data.

Reply:
Fixed by also going to chain last pbft block if queue is empty